campaigns: add GitLab webhook support

LawnGnome opened the pull request:
This PR closes #12171 by implementing GitLab webhook support. It includes the UI and schema changes required for GitLab external services to support webhooks, the documentation for a user to configure the services, and then all the bits we need to handle those webhook requests.

Random points of interest are enumerated below.

### Testing

About half of this PR is unit and integration tests for the new `webhooks_gitlab.go` file, which I've put a lot of effort into testing as thoroughly as possible, since we don't control what comes into the HTTP handler. To shamelessly rip off one of [Knuth's great lines](https://www-cs-faculty.stanford.edu/~knuth/faq.html), I can't promise that it's bug free, but I can promise it has 100% code coverage.

### Approvals and unapprovals

GitLab uses a completely different mechanism in webhooks to indicate that a merge request has been approved or unapproved to what the API normally exposes. In the REST and GraphQL APIs, we get the approval state by parsing the list of notes (comments) that are attached to the merge request: [there's a concept of a "system note" that we can string parse to figure out the state](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/repo-updater/repos/gitlab.go#L467-469). (Alas, more structured approval and unapproval data relies on the project (a) using an enterprise licence, and (b) being configured with approval pipelines, which isn't required for normal use.)

In webhooks, we just get the merge request as a "merge request event" with an "action" field set to "approved" or "unapproved", and no further details. (I'd point to the documentation, but I couldn't find any. This was all trial and error.) The lack of detail means we can't deduplicate the approval later when we do a full sync, since we don't have the note ID. To make matters worse, although GitLab does support note events, note events aren't sent for system notes.

So I've taken a slightly different approach here: we don't create changeset events based on approval or unapproval webhook invocations. Instead, we take those as a hint that we should sync the merge request via `repo-updater` as soon as possible, and let the normal sync process handle it as if the user had clicked that lovely refresh button in the UI.

Other merge request events are handled more normally.

### Time

GitLab's REST and GraphQL APIs use RFC 3339 dates. As I'm sure you all know, [Go can parse these directly out of JSON into `time.Time` instances](https://golang.org/pkg/time/#Time.UnmarshalJSON).

GitLab's webhook events are... less consistent. Some fields are RFC 3339, whereas others [are not](https://gitlab.com/gitlab-org/gitlab/-/issues/19567).

As a result, I've updated the structures in the GitLab package to use a new `gitlab.Time` instead of `time.Time` that can handle them both. There's a bit of extra boilerplate here around type assertions and casts, but it means we can then use the same structures for types that are shared between the REST and webhook APIs, so I'm calling it a win.

### Pipelines

Pipeline events [are supposed to include the merge request they were triggered by](https://docs.gitlab.com/ee/user/project/integrations/webhooks.html#pipeline-events). [In practice, they generally do not.](https://gitlab.com/gitlab-org/gitlab/-/issues/208686)

I've speculatively implemented pipeline event support based on the GitLab docs in the hopes that the bug will eventually be addressed. For now, however, I've been unable to actually trigger it in practice because I haven't been able to generate a pipeline event that includes the merge request. Again, we'll fall back to `repo-updater` figuring this out when it syncs, since we _can_ get the pipeline events through the REST (and, eventually, GraphQL) API.

LawnGnome commented:
This is ready for review.

eseliger commented:
> I've speculatively implemented pipeline event support based on the GitLab docs in the hopes that the bug will eventually be addressed. For now, however, I've been unable to actually trigger it in practice because I haven't been able to generate a pipeline event that includes the merge request. Again, we'll fall back to repo-updater figuring this out when it syncs, since we can get the pipeline events through the REST (and, eventually, GraphQL) API.

The linked issue has been closed as it was wrongly identified as a support request. Do you think we should take action here (open a new one or so, and track it?)

eseliger commented on enterprise/internal/campaigns/webhooks_gitlab_test.go:
```diff
@@ -0,0 +1,1008 @@
+						}
+						req.Header.Add(webhooks.TokenHeaderName, "secret")
+
+						repoupdater.MockEnqueueChangesetSync = func(ctx context.Context, ids []int64) error {
+							if diff := cmp.Diff(ids, []int64{changeset.ID}); diff != "" {
```
should we also assert this function is called?

```suggestion
                                                  changesetEnqueued := false
						repoupdater.MockEnqueueChangesetSync = func(ctx context.Context, ids []int64) error {
				                         changesetEnqueued = true
							if diff := cmp.Diff(ids, []int64{changeset.ID}); diff != "" {
```

LawnGnome replied:
Yes. Yes, we should.

mrnugget commented on enterprise/internal/campaigns/webhooks_gitlab_test.go:
```diff
@@ -0,0 +1,1008 @@
+// It would be theoretically possible to use savepoints to implement something
+// resembling the semantics of a true nested transaction, but that's
+// unnecessary for these tests.
+type nestedTx struct{ *sql.Tx }
```
Nice! In other tests (see `integration_test.go`) we specifically create a new test database, because we can't create transactions in the existing test-transaction.

The new `basestore` package can handle savepoints though: https://github.com/sourcegraph/sourcegraph/blob/97ed62cbeb800072dc04b08ab5280e07d3b9a23a/internal/db/basestore/handle.go#L42-L50

eseliger commented on doc/admin/external_service/gitlab.md:
```diff
@@ -84,3 +84,32 @@ We are actively collaborating with GitLab to improve our integration (e.g. the [
+- Merge request events
+- Pipeline events
+
+Webhooks must be configured on each project that you wish to monitor on GitLab. To set up a project webhook on GitLab, go to the Webhook settings page of the project.
```
It took me a while to realize the "Integrations" settings page is the "webhooks" settings page, should we clarify that?

LawnGnome replied:
Did you look at gitlab.com, or a private instance? On the former, it's a distinct page in the navigation:

![image](https://user-images.githubusercontent.com/229984/88742018-e70a5380-d0f5-11ea-9270-578f58565ed4.png)

Happy to clarify the wording, but I'm wondering if there's a version difference here somehow.

LawnGnome replied:
(reposting with a better cropped/drawn screenshot)

On gitlab.com, "webhooks" is its own settings page:

![image](https://user-images.githubusercontent.com/229984/88742159-584a0680-d0f6-11ea-9cd2-43978cc8b637.png)

Did you see something different elsewhere? I wonder if there's a version difference here.

eseliger replied:
Oh! Indeed on our gitlab.sgdev.org it looks like this:

![image](https://user-images.githubusercontent.com/19534377/88742266-18a90c00-d142-11ea-9a0b-d3217de236a3.png)

So this is probably us having a very dated instance(?)

LawnGnome replied:
Thanks! I might play around with the wording a touch to indicate that, then.

mrnugget commented on enterprise/internal/campaigns/webhooks_gitlab_test.go:
```diff
@@ -0,0 +1,1008 @@
+			})
+		})
+
+		t.Run("getExternalServiceFromRawID database error", func(t *testing.T) {
+			// This test is separate from the other unit tests for this
+			// function above because it needs to set up a bad database
+			// connection on the repo store.
+			store, _, clock := gitLabTestSetup(t, db)
+			rstore := repos.NewDBStore(&brokenDB{errors.New("foo")}, sql.TxOptions{})
+			h := NewGitLabWebhook(store, rstore, clock.now)
+
+			_, err := h.getExternalServiceFromRawID(ctx, "12345")
+			if err == nil {
+				t.Error("unexpected nil error")
+			}
+		})
+
+		t.Run("database error", func(t *testing.T) {
+			// We can induce an error with a broken database connection.
+			store, rstore, clock := gitLabTestSetup(t, db)
+			h := NewGitLabWebhook(store, rstore, clock.now)
+			h.Store = NewStoreWithClock(&brokenDB{errors.New("foo")}, clock.now)
+
+			es, err := h.getExternalServiceFromRawID(ctx, "12345")
+			if es != nil {
+				t.Errorf("unexpected non-nil external service: %+v", es)
+			}
+			if err == nil {
+				t.Error("unexpected nil error")
+			}
+		})
```
Tiny nitpicky suggestion, feel free to ignore: these tests test the same thing, except one has a broken `campaigns.Store` and the other a broken `repos.Store`, right? If the name passed to `t.Run` would follow the same schema and the assertions would be the same, that would be clearer, I guess.

LawnGnome replied:
You can definitely tell one of those test functions came well after the other. Updated.

eseliger commented:
## Manual testing report

- Integrations settings page was misleading, maybe we should help in the docs, also added that comment inline above
- All not supported events log errors, but they are frequent, I think we should just ignore them (or log in verbose) and also not fail the request since GitLab might retry.
- Closing yielded 204 but didn’t update the changeset in the UI (changeset event exists) - a resync works though
- Pending pipeline is not detected, stays in state “UNKNOWN” - it’s auto devops, maybe that’s the reason why? - Update: appeared after a manual sync, but was out of sync even though webhooks were enabled, not sure that's detectable(?)
- Shows check state PENDING although no pipeline exists - also no auto devops pipeline
- Labelling events don’t work (501) - only appear after resync. I saw you added a label type that has the color set, does that mean we can go away from the hardcoded `#000000`?
- Our Gitlab instance appears to not support reviews - so couldn’t test them
- Panic caused by close event, after first closed, then reopened and then the second close (and reopen also fails now). So basically from the third merge state webhook on, it fails.
- Cancelling the pipeline triggers an event which is successful (204), but the UI still shows pending, that was unexpected, but I’m not sure this is a fault of this PR. How do we handle cancelled otherwise? Failed? Unknown?
- After merge it appeared closed in the UI, resync fixed.
- It doesn’t appear as merged in the changesetCountsOverTime resolver (still pending) (perhaps also other states not reflected, didn't test deeper here)

Panic:
```
18:26:54                          frontend | goroutine 240291 [running]:
18:26:54                          frontend | runtime/debug.Stack(0xc0002a7888, 0x5a82f20, 0xc003204ae0)
18:26:54                          frontend |    /usr/local/Cellar/go/1.14.2_1/libexec/src/runtime/debug/stack.go:24 +0x9d
18:26:54                          frontend | runtime/debug.PrintStack()
18:26:54                          frontend |    /usr/local/Cellar/go/1.14.2_1/libexec/src/runtime/debug/stack.go:16 +0x22
18:26:54                          frontend | github.com/getsentry/raven-go.Recoverer.func1.1(0xc003136400, 0x5ec8280, 0xc00031fce0)
18:26:54                          frontend |    /Users/erik/Code/go/pkg/mod/github.com/getsentry/raven-go@v0.2.0/http.go:84 +0x6e
18:26:54                          frontend | panic(0x55d7e80, 0xc0020561c0)
18:26:54                          frontend |    /usr/local/Cellar/go/1.14.2_1/libexec/src/runtime/panic.go:969 +0x166
18:26:54                          frontend | github.com/sourcegraph/sourcegraph/internal/campaigns.(*ChangesetEvent).Update(0xc0015fe9a0, 0xc0015fe930)
18:26:54                          frontend |    /Users/erik/Code/sourcegraph/sourcegraph/internal/campaigns/types.go:1206 +0x3bb3
18:26:54                          frontend | github.com/sourcegraph/sourcegraph/enterprise/internal/campaigns.Webhook.upsertChangesetEvent(0xc0004774d0, 0x5edc9e0, 0xc000321ae0, 0x5a82350, 0x58807f7, 0x6, 0x5ed1ac0, 0xc00189c7b0, 0xc00159e880, 0x19, ...)
18:26:54                          frontend |    /Users/erik/Code/sourcegraph/sourcegraph/enterprise/internal/campaigns/webhooks.go:156 +0x979
18:26:54                          frontend | github.com/sourcegraph/sourcegraph/enterprise/internal/campaigns.(*GitLabWebhook).handleMergeRequestStateEvent(0xc001b8c608, 0x5ed1ac0, 0xc00189c7b0, 0xc00159e880, 0x19, 0xf8c2030, 0xc000568480, 0x0, 0x0)
18:26:54                          frontend |    /Users/erik/Code/sourcegraph/sourcegraph/enterprise/internal/campaigns/webhooks_gitlab.go:215 +0x194
18:26:54                          frontend | github.com/sourcegraph/sourcegraph/enterprise/internal/campaigns.(*GitLabWebhook).handleEvent(0xc001b8c608, 0x5ed1ac0, 0xc00189c7b0, 0xc0027bb380, 0x55da580, 0xc000568480, 0x0)
18:26:54                          frontend |    /Users/erik/Code/sourcegraph/sourcegraph/enterprise/internal/campaigns/webhooks_gitlab.go:145 +0x5bc
18:26:54                          frontend | github.com/sourcegraph/sourcegraph/enterprise/internal/campaigns.(*GitLabWebhook).ServeHTTP(0xc001b8c608, 0x5ec9f40, 0xc0033b4cd0, 0xc003136800)
18:26:54                          frontend |    /Users/erik/Code/sourcegraph/sourcegraph/enterprise/internal/campaigns/webhooks_gitlab.go:73 +0x2f5
18:26:54                          frontend | github.com/sourcegraph/sourcegraph/internal/trace.TraceRoute.func1(0x5ec9f40, 0xc0033b4cd0, 0xc003136800)
18:26:54                          frontend |    /Users/erik/Code/sourcegraph/sourcegraph/internal/trace/httptrace.go:264 +0xbc
18:26:54                          frontend | net/http.HandlerFunc.ServeHTTP(0xc00225b580, 0x5ec9f40, 0xc0033b4cd0, 0xc003136800)
18:26:54                          frontend |    /usr/local/Cellar/go/1.14.2_1/libexec/src/net/http/server.go:2012 +0x44
18:26:54                          frontend | github.com/gorilla/mux.(*Router).ServeHTTP(0xc002264b40, 0x5ec9f40, 0xc0033b4cd0, 0xc003136600)
18:26:54                          frontend |    /Users/erik/Code/go/pkg/mod/github.com/gorilla/mux@v1.7.4/mux.go:210 +0xe2
18:26:54                          frontend | github.com/sourcegraph/sourcegraph/enterprise/cmd/frontend/authz.Init.func3.1(0x5ec9f40, 0xc0033b4cd0, 0xc003136600)
18:26:54                          frontend |    /Users/erik/Code/sourcegraph/sourcegraph/enterprise/cmd/frontend/authz/authz.go:118 +0x429
18:26:54                          frontend | net/http.HandlerFunc.ServeHTTP(0xc00225b6e0, 0x5ec9f40, 0xc0033b4cd0, 0xc003136600)
18:26:54                          frontend |    /usr/local/Cellar/go/1.14.2_1/libexec/src/net/http/server.go:2012 +0x44
18:26:54                          frontend | github.com/sourcegraph/sourcegraph/cmd/frontend/auth.glob..func1.1(0x5ec9f40, 0xc0033b4cd0, 0xc003136600)
18:26:54                          frontend |    /Users/erik/Code/sourcegraph/sourcegraph/cmd/frontend/auth/non_public.go:35 +0x104
18:26:54                          frontend | net/http.HandlerFunc.ServeHTTP(0xc00225b700, 0x5ec9f40, 0xc0033b4cd0, 0xc003136600)
18:26:54                          frontend |    /usr/local/Cellar/go/1.14.2_1/libexec/src/net/http/server.go:2012 +0x44
18:26:54                          frontend | github.com/sourcegraph/sourcegraph/enterprise/cmd/frontend/auth/openidconnect.handleOpenIDConnectAuth(0x5ec9f40, 0xc0033b4cd0, 0xc003136600, 0x5e9ae60, 0xc00225b700, 0x1)
18:26:54                          frontend |    /Users/erik/Code/sourcegraph/sourcegraph/enterprise/cmd/frontend/auth/openidconnect/middleware.go:101 +0x149
18:26:54                          frontend | github.com/sourcegraph/sourcegraph/enterprise/cmd/frontend/auth/openidconnect.glob..func1.1(0x5ec9f40, 0xc0033b4cd0, 0xc003136600)
18:26:54                          frontend |    /Users/erik/Code/sourcegraph/sourcegraph/enterprise/cmd/frontend/auth/openidconnect/middleware.go:53 +0x56
18:26:54                          frontend | net/http.HandlerFunc.ServeHTTP(0xc00225b720, 0x5ec9f40, 0xc0033b4cd0, 0xc003136600)
18:26:54                          frontend |    /usr/local/Cellar/go/1.14.2_1/libexec/src/net/http/server.go:2012 +0x44
18:26:54                          frontend | github.com/sourcegraph/sourcegraph/enterprise/cmd/frontend/auth/saml.authHandler(0x5ec9f40, 0xc0033b4cd0, 0xc003136600, 0x5e9ae60, 0xc00225b720, 0x1)
18:26:54                          frontend |    /Users/erik/Code/sourcegraph/sourcegraph/enterprise/cmd/frontend/auth/saml/middleware.go:69 +0xf4
18:26:54                          frontend | github.com/sourcegraph/sourcegraph/enterprise/cmd/frontend/auth/saml.glob..func1.1(0x5ec9f40, 0xc0033b4cd0, 0xc003136600)
18:26:54                          frontend |    /Users/erik/Code/sourcegraph/sourcegraph/enterprise/cmd/frontend/auth/saml/middleware.go:30 +0x56
18:26:54                          frontend | net/http.HandlerFunc.ServeHTTP(0xc00225b740, 0x5ec9f40, 0xc0033b4cd0, 0xc003136600)
18:26:54                          frontend |    /usr/local/Cellar/go/1.14.2_1/libexec/src/net/http/server.go:2012 +0x44
18:26:54                          frontend | github.com/sourcegraph/sourcegraph/enterprise/cmd/frontend/auth/httpheader.middleware.func1(0x5ec9f40, 0xc0033b4cd0, 0xc003136600)
18:26:54                          frontend |    /Users/erik/Code/sourcegraph/sourcegraph/enterprise/cmd/frontend/auth/httpheader/middleware.go:46 +0xad7
18:26:54                          frontend | net/http.HandlerFunc.ServeHTTP(0xc00225b760, 0x5ec9f40, 0xc0033b4cd0, 0xc003136600)
18:26:54                          frontend |    /usr/local/Cellar/go/1.14.2_1/libexec/src/net/http/server.go:2012 +0x44
18:26:54                          frontend | github.com/sourcegraph/sourcegraph/enterprise/cmd/frontend/auth/oauth.NewHandler.func1(0x5ec9f40, 0xc0033b4cd0, 0xc003136600)
18:26:54                          frontend |    /Users/erik/Code/sourcegraph/sourcegraph/enterprise/cmd/frontend/auth/oauth/middleware.go:43 +0x433
18:26:54                          frontend | net/http.HandlerFunc.ServeHTTP(0xc0000b2c80, 0x5ec9f40, 0xc0033b4cd0, 0xc003136600)
18:26:54                          frontend |    /usr/local/Cellar/go/1.14.2_1/libexec/src/net/http/server.go:2012 +0x44
18:26:54                          frontend | github.com/sourcegraph/sourcegraph/enterprise/cmd/frontend/auth/oauth.NewHandler.func1(0x5ec9f40, 0xc0033b4cd0, 0xc003136600)
18:26:54                          frontend |    /Users/erik/Code/sourcegraph/sourcegraph/enterprise/cmd/frontend/auth/oauth/middleware.go:43 +0x433
18:26:54                          frontend | net/http.HandlerFunc.ServeHTTP(0xc0000b2d00, 0x5ec9f40, 0xc0033b4cd0, 0xc003136600)
```

Body:
```json
{
    "object_kind": "merge_request",
    "event_type": "merge_request",
    "user": {
        "name": "Erik Seliger",
        "username": "eseliger",
        "avatar_url": "https://secure.gravatar.com/avatar/12bba0958b79429975b4f9a41d709ece?s=80&d=identicon"
    },
    "project": {
        "id": 4074,
        "name": "bitbucket-server-plugin",
        "description": "Get code intelligence and run Sourcegraph extensions on Bitbucket Server.",
        "web_url": "https://gitlab.sgdev.org/sourcegraph/bitbucket-server-plugin",
        "avatar_url": null,
        "git_ssh_url": "git@gitlab.sgdev.org:sourcegraph/bitbucket-server-plugin.git",
        "git_http_url": "https://gitlab.sgdev.org/sourcegraph/bitbucket-server-plugin.git",
        "namespace": "sourcegraph",
        "visibility_level": 10,
        "path_with_namespace": "sourcegraph/bitbucket-server-plugin",
        "default_branch": "master",
        "ci_config_path": null,
        "homepage": "https://gitlab.sgdev.org/sourcegraph/bitbucket-server-plugin",
        "url": "git@gitlab.sgdev.org:sourcegraph/bitbucket-server-plugin.git",
        "ssh_url": "git@gitlab.sgdev.org:sourcegraph/bitbucket-server-plugin.git",
        "http_url": "https://gitlab.sgdev.org/sourcegraph/bitbucket-server-plugin.git"
    },
    "object_attributes": {
        "assignee_id": null,
        "author_id": 16,
        "created_at": "2020-02-19 14:20:07 UTC",
        "description": "*Created by: kzh*\r\n\r\nThis PR adds a minimal release script. It produces the plugin jar and uploads it to our google cloud storage bucket.\r\n\r\nIt's a WIP. Git tagging will likely be added. The release process isn't determined at the moment. I'm also considering doing all of this in a CI process.\r\n",
        "head_pipeline_id": 55,
        "id": 2556,
        "iid": 19,
        "last_edited_at": "2020-07-28 16:15:01 UTC",
        "last_edited_by_id": 31,
        "merge_commit_sha": null,
        "merge_error": null,
        "merge_params": {
            "force_remove_source_branch": "0"
        },
        "merge_status": "can_be_merged",
        "merge_user_id": null,
        "merge_when_pipeline_succeeds": false,
        "milestone_id": null,
        "source_branch": "kz/release-scripts",
        "source_project_id": 4074,
        "state_id": 2,
        "target_branch": "master",
        "target_project_id": 4074,
        "time_estimate": 0,
        "title": "Add release script",
        "updated_at": "2020-07-28 16:26:53 UTC",
        "updated_by_id": 31,
        "url": "https://gitlab.sgdev.org/sourcegraph/bitbucket-server-plugin/merge_requests/19",
        "source": {
            "id": 4074,
            "name": "bitbucket-server-plugin",
            "description": "Get code intelligence and run Sourcegraph extensions on Bitbucket Server.",
            "web_url": "https://gitlab.sgdev.org/sourcegraph/bitbucket-server-plugin",
            "avatar_url": null,
            "git_ssh_url": "git@gitlab.sgdev.org:sourcegraph/bitbucket-server-plugin.git",
            "git_http_url": "https://gitlab.sgdev.org/sourcegraph/bitbucket-server-plugin.git",
            "namespace": "sourcegraph",
            "visibility_level": 10,
            "path_with_namespace": "sourcegraph/bitbucket-server-plugin",
            "default_branch": "master",
            "ci_config_path": null,
            "homepage": "https://gitlab.sgdev.org/sourcegraph/bitbucket-server-plugin",
            "url": "git@gitlab.sgdev.org:sourcegraph/bitbucket-server-plugin.git",
            "ssh_url": "git@gitlab.sgdev.org:sourcegraph/bitbucket-server-plugin.git",
            "http_url": "https://gitlab.sgdev.org/sourcegraph/bitbucket-server-plugin.git"
        },
        "target": {
            "id": 4074,
            "name": "bitbucket-server-plugin",
            "description": "Get code intelligence and run Sourcegraph extensions on Bitbucket Server.",
            "web_url": "https://gitlab.sgdev.org/sourcegraph/bitbucket-server-plugin",
            "avatar_url": null,
            "git_ssh_url": "git@gitlab.sgdev.org:sourcegraph/bitbucket-server-plugin.git",
            "git_http_url": "https://gitlab.sgdev.org/sourcegraph/bitbucket-server-plugin.git",
            "namespace": "sourcegraph",
            "visibility_level": 10,
            "path_with_namespace": "sourcegraph/bitbucket-server-plugin",
            "default_branch": "master",
            "ci_config_path": null,
            "homepage": "https://gitlab.sgdev.org/sourcegraph/bitbucket-server-plugin",
            "url": "git@gitlab.sgdev.org:sourcegraph/bitbucket-server-plugin.git",
            "ssh_url": "git@gitlab.sgdev.org:sourcegraph/bitbucket-server-plugin.git",
            "http_url": "https://gitlab.sgdev.org/sourcegraph/bitbucket-server-plugin.git"
        },
        "last_commit": {
            "id": "c00d24d2b0805292f74679b44928f5d20e84166c",
            "message": "Merge branch 'master' into 'kz/release-scripts'\n\n# Conflicts:\n#   README.md",
            "timestamp": "2020-07-28T16:12:57+00:00",
            "url": "https://gitlab.sgdev.org/sourcegraph/bitbucket-server-plugin/commit/c00d24d2b0805292f74679b44928f5d20e84166c",
            "author": {
                "name": "Erik Seliger",
                "email": "erik@sourcegraph.com"
            }
        },
        "work_in_progress": false,
        "total_time_spent": 0,
        "human_total_time_spent": null,
        "human_time_estimate": null,
        "assignee_ids": [],
        "state": "closed",
        "action": "close"
    },
    "labels": [
        {
            "id": 315,
            "title": "bug",
            "color": "#d73a4a",
            "project_id": 4074,
            "created_at": "2020-03-06 09:59:31 UTC",
            "updated_at": "2020-03-06 09:59:31 UTC",
            "template": false,
            "description": null,
            "type": "ProjectLabel",
            "group_id": null
        }
    ],
    "changes": {
        "state_id": {
            "previous": 1,
            "current": 2
        },
        "updated_at": {
            "previous": "2020-07-28 16:22:10 UTC",
            "current": "2020-07-28 16:26:53 UTC"
        },
        "labels": {
            "previous": [],
            "current": [
                {
                    "id": 315,
                    "title": "bug",
                    "color": "#d73a4a",
                    "project_id": 4074,
                    "created_at": "2020-03-06 09:59:31 UTC",
                    "updated_at": "2020-03-06 09:59:31 UTC",
                    "template": false,
                    "description": null,
                    "type": "ProjectLabel",
                    "group_id": null
                }
            ]
        },
        "total_time_spent": {
            "previous": null,
            "current": 0
        }
    },
    "repository": {
        "name": "bitbucket-server-plugin",
        "url": "git@gitlab.sgdev.org:sourcegraph/bitbucket-server-plugin.git",
        "description": "Get code intelligence and run Sourcegraph extensions on Bitbucket Server.",
        "homepage": "https://gitlab.sgdev.org/sourcegraph/bitbucket-server-plugin"
    }
}
```

eseliger commented on internal/extsvc/gitlab/webhooks/merge_requests.go:
```diff
@@ -0,0 +1,94 @@
+	return fmt.Sprintf("MergeRequest:%s:%d:%s", prefix, e.MergeRequest.IID, e.MergeRequest.CreatedAt.Format(time.RFC3339))
+}
+
+type mergeRequestEvent struct {
```
I'm with you

LawnGnome replied:
I've added some documentation, and renamed `MergeRequestEvent` to `MergeRequestEventCommon`. Please let me know if it's still unclear.

LawnGnome commented on internal/extsvc/gitlab/webhooks/events.go:
```diff
@@ -0,0 +1,78 @@
+}
+
+// Simple events that are simply unmarshalled and have no methods attached are defined below.
+
```
I actually did this intentionally, since it's not a docblock for `PipelineEvent` as such. (But, equally, there used to more types than just `PipelineEvent` under this comment.)

Do you think I should change it to a more specific docblock?

eseliger replied:
Ah I didn't notice the plural 🤔 All good then!

LawnGnome commented on internal/extsvc/gitlab/webhooks/merge_requests.go:
```diff
@@ -0,0 +1,94 @@
+func (e *MergeRequestReopenEvent) Key() string { return e.key("Reopen") }
+
+func (e *MergeRequestEvent) key(prefix string) string {
+	// We can't key solely off the merge request ID because it may be reopened
+	// and closed multiple times. Instead, we'll use the CreatedAt field.
+	return fmt.Sprintf("MergeRequest:%s:%d:%s", prefix, e.MergeRequest.IID, e.MergeRequest.CreatedAt.Format(time.RFC3339))
```
Yes, it should be the event. Good catch!

LawnGnome replied:
Actually, I take that back: it was meant to be the updated time. But it was still wrong. :smiley:

LawnGnome commented on enterprise/internal/campaigns/webhooks_gitlab.go:
```diff
@@ -0,0 +1,285 @@
+	event, err := webhooks.UnmarshalEvent(payload)
+	if err != nil {
+		if errors.Is(err, webhooks.ErrObjectKindUnknown) {
+			respond(w, http.StatusNotImplemented, err)
```
Yeah, that's a good point. I'll make the relevant updates for this and the below comment.

LawnGnome commented on enterprise/internal/campaigns/webhooks_gitlab.go:
```diff
@@ -0,0 +1,285 @@
+	}
+
+	for _, e := range es {
+		if e.ID == id {
```
That was indeed why: I wasn't sure if we had some sort of expectation that `externalServiceID` be optional across code hosts. I'll remove the loop.

LawnGnome commented on internal/extsvc/gitlab/webhooks/merge_requests.go:
```diff
@@ -0,0 +1,94 @@
+
+func (e *MergeRequestCloseEvent) Key() string  { return e.key("Close") }
+func (e *MergeRequestMergeEvent) Key() string  { return e.key("Merge") }
+func (e *MergeRequestReopenEvent) Key() string { return e.key("Reopen") }
```
Added.

LawnGnome commented on web/src/site-admin/SiteAdminExternalServiceWebhook.tsx:
```diff
@@ -0,0 +1,92 @@
+export const SiteAdminExternalServiceWebhook: React.FunctionComponent<Props> = props => {
+    const { kind, webhookURL } = props.externalService
+
+    if (webhookURL) {
```
Done. :smiley:

LawnGnome commented:
Just handling the things that weren't the main bug in @eseliger's excellent testing comment (thanks!):

> * Integrations settings page was misleading, maybe we should help in the docs, also added that comment inline above

Updated in 6b3cf11eb4be17e3d77b5aaa8874cf40c7eb1226.

> * All not supported events log errors, but they are frequent, I think we should just ignore them (or log in verbose) and also not fail the request since GitLab might retry.

Return values changed in 7e7753a2aa1e652038f3dfe0bd8c8aa987f6939c, but they're at `info` right now. Happy to change this if we think users are likely to configure event types that we can't handle.

> * Pending pipeline is not detected, stays in state “UNKNOWN” - it’s auto devops, maybe that’s the reason why? - Update: appeared after a manual sync, but was out of sync even though webhooks were enabled, not sure that's detectable(?)

This is the pipelines not including merge request issue noted in the initial PR description. I concur with your earlier inline comment that the GitLab issue should be reopened, and I'll do that when I have a good example to hand to provide.

> * Labelling events don’t work (501) - only appear after resync. I saw you added a label type that has the color set, does that mean we can go away from the hardcoded `#000000`?

Not yet. For consistency with syncs (where we don't get the colour), I haven't wired this up in this PR. When we switch to GraphQL, we can wire this up, but I figured we may as well unmarshal the fields now.

> * Cancelling the pipeline triggers an event which is successful (204), but the UI still shows pending, that was unexpected, but I’m not sure this is a fault of this PR. How do we handle cancelled otherwise? Failed? Unknown?
>
> * After merge it appeared closed in the UI, resync fixed.

Again, this would be the issue we have where we can't tie a pipeline event back to a specific merge request. It's annoying, to say the least.

Things I still have to investigate:

> * Closing yielded 204 but didn’t update the changeset in the UI (changeset event exists) - a resync works though

Need to investigate this. Could be related to @mrnugget's questions about how we calculate derived state.

> * Shows check state PENDING although no pipeline exists - also no auto devops pipeline

I think I know why this is happening; I don't (yet) know if we can detect it.
 
> * Panic caused by close event, after first closed, then reopened and then the second close (and reopen also fails now). So basically from the third merge state webhook on, it fails.

Will investigate.

> * It doesn’t appear as merged in the changesetCountsOverTime resolver (still pending) (perhaps also other states not reflected, didn't test deeper here)

Will investigate.

LawnGnome commented on doc/admin/external_service/gitlab.md:
```diff
@@ -84,3 +84,32 @@ We are actively collaborating with GitLab to improve our integration (e.g. the [
+
+Click on **Enable SSL verification** if you have configured SSL with a valid certificate in your Sourcegraph instance.
+
+Finally, select **Add webhook** to create the webhook.
```
Done for GitLab in this PR; opened #12529 for the other code hosts.

LawnGnome commented:
> * Panic caused by close event, after first closed, then reopened and then the second close (and reopen also fails now). So basically from the third merge state webhook on, it fails.

This should be fixed by b39e54ed32b5674f4956815cfe86650e0e8180a5.

The underlying panic was also arguably a bug in our update code; 890f978dd0bfb9eac2c5d7a7d256300d2ec57de2 addresses that by at least turning it into an error we can report normally.

> * It doesn’t appear as merged in the changesetCountsOverTime resolver (still pending) (perhaps also other states not reflected, didn't test deeper here)

I fixed this one too, but I'm seeing some unreliability in how we invoke `SetDerivedState()`, specifically around the timestamp-based checks the compute functions (such as `computeExternalState()`) use to try to avoid iterating the full set of events.

The thing is: we _already_ iterate over the full set of changeset events in `computeHistory()`. So I think we've actually overcomplicated this: as long as we synthesise the appropriate events on sync (which we have to anyway for the burndown chart), we could compute the external states we care about while we also calculate the history and get rid of `computeExternalState()` and `computeReviewState()` entirely.

I'm going to sketch this out (in a different PR) tomorrow morning.

mrnugget commented:
> The thing is: we _already_ iterate over the full set of changeset events in `computeHistory()`. So I think we've actually overcomplicated this: as long as we synthesise the appropriate events on sync (which we have to anyway for the burndown chart), we could compute the external states we care about while we also calculate the history and get rid of `computeExternalState()` and `computeReviewState()` entirely.

Yeah, there's a ton of possible optimisations in `computeHistory`. Right now it's loops in loops and we could be a lot smarter about that, `computeHistory` itself _is_ an improvement on what was there before, but, well :) 

Just a word of caution though: before you fall into an optimisation rabbit hole, I'd pull back, create a ticket and revisit once we have the new workflow working.

mrnugget commented:
Before I forget it again: add an entry to the `CHANGELOG.md` 😃

LawnGnome commented:
> Just a word of caution though: before you fall into an optimisation rabbit hole, I'd pull back, create a ticket and revisit once we have the new workflow working.

I agree (as we talked about in our buddy sync). Filed #12565.

LawnGnome commented:
It took a bit to verify it, but I believe b39e54ed32b5674f4956815cfe86650e0e8180a5 fixed the remaining issues @eseliger found as well.

Which leaves only this question from @mrnugget:

> * Do we have a bug in `syncChangesetsWithSources` and we need to first insert the new changeset events, then load _all_ events again, and _then_ set the derived state?
> 
> * Did you know all of this and had this in mind? smile

In summary, yes. I do think we have some ordering issues, and I think fixing #12565 will fix those, but I've tried to work around them as best I can in here.

Therefore: I think this is ready for final testing/review.

eseliger commented on enterprise/internal/campaigns/webhooks_gitlab.go:
```diff
@@ -0,0 +1,292 @@
+	// to avoid unexpected retries.
+	//
+	// To meet this goal, rather than synchronously synchronizing here, we'll
+	// instead ask repo-updater to prioritise the sync of this changeset and let
```
```suggestion
	// instead ask repo-updater to prioritize the sync of this changeset and let
```

LawnGnome commented:
> Unknown events are still logged as errors - but they can happen, think milestone assignment events or so, they're covered in merge request events but not supported by us. I think they should be marked less harmful

I fixed that one up in 8d1e6df. I agree it was too noisy.

> Label events are still not handled - we do that for GitHub though. Doesn’t need to take the color into account yet, but I think feature parity for webhooks would be great, so it’s less documentation work why certain things aren’t real time while others are. Not sure I understood your comment regarding those events correctly, but I think you only referred to color support right?

Sadly, GitLab does not give us "label" events per se: we get a merge request update event, along with a changes field that is unspecified in nature but _might_ include some labels if the labels change. Rather than trying to pull that apart, I've made update events do the same thing as approval/unapproval: ask `repo-updater` to do a sync ASAP, which also means we'll have faster updates for other changes.

That's db3a23e for the unmarshalling, and fc6f1a9 for the event handling.

> https://gitlab.sgdev.org/sourcegraph/bitbucket-server-plugin/merge_requests/22 has no pipeline, also no devops one but check state is PENDING

I can't reproduce this one, nor can I figure out how the code would have gotten to pending: looking at the return value from the REST API, there's no pipeline (as you'd expect), so the relevant code should short circuit out to `UNKNOWN`, which is (as far as I know) correct.

Do you have the database floating around for that changeset? I'd love to see the changeset record and any events associated with it.

unknwon commented on internal/extsvc/gitlab/webhooks/events.go:
```diff
@@ -0,0 +1,78 @@
+package webhooks
```
qq: what's the rationale to make a new subpackage? 🤔

LawnGnome replied:
It felt logically separate to me: while there's overlap between GitLab's REST API and webhooks (the types have most, but not all, of their fields in common, and the webhooks package reuses those types where appropriate), the webhook API also has plenty of its own quirks, and I preferred to isolate the behaviour we needed around typing and parsing webhook objects outside of the general gitlab package to avoid clutter and make it easier to reason about and test.

eseliger commented:
Great, thanks for fixing the remaining things! 

Yeah I still have it in my database, and for some spurious reason, there actually is _one_ event in the DB:

<details>
<summary>Changeset row</summary>

1 | {"1": null} | 253 | 2020-07-28 17:34:45.242012+02 | 2020-07-31 14:32:43.699221+02 | {"id": 13696, "iid": 22, "Notes": [{"id": 35203, "body": "mentioned in commit 08080ac8c1e6dd2e02c777918350fa9d005f1bfc", "author": {"id": 31, "name": "Erik Seliger", "email": "", "state": "active", "web_url": "https://gitlab.sgdev.org/eseliger", "username": "eseliger", "avatar_url": "https://secure.gravatar.com/avatar/12bba0958b79429975b4f9a41d709ece?s=80&d=identicon", "identities": null}, "system": true, "created_at": "2020-07-31T11:48:23.988Z"}, {"id": 35202, "body": "merged", "author": {"id": 31, "name": "Erik Seliger", "email": "", "state": "active", "web_url": "https://gitlab.sgdev.org/eseliger", "username": "eseliger", "avatar_url": "https://secure.gravatar.com/avatar/12bba0958b79429975b4f9a41d709ece?s=80&d=identicon", "identities": null}, "system": true, "created_at": "2020-07-31T11:48:23.521Z"}, {"id": 35201, "body": "added 15 commits\n\n<ul><li>0df1ddfa...04408927 - 14 commits from branch <code>master</code></li><li>d3647d13 - Merge branch &#39;master&#39; into &#39;kz/upsert-webhooks&#39;</li></ul>\n\n[Compare with previous version](/sourcegraph/bitbucket-server-plugin/merge_requests/22/diffs?diff_id=13683&start_sha=0df1ddfaaee36e74cb4fea8b625b30475a78df8c)", "author": {"id": 31, "name": "Erik Seliger", "email": "", "state": "active", "web_url": "https://gitlab.sgdev.org/eseliger", "username": "eseliger", "avatar_url": "https://secure.gravatar.com/avatar/12bba0958b79429975b4f9a41d709ece?s=80&d=identicon", "identities": null}, "system": true, "created_at": "2020-07-31T11:44:33.388Z"}, {"id": 35200, "body": "reopened", "author": {"id": 31, "name": "Erik Seliger", "email": "", "state": "active", "web_url": "https://gitlab.sgdev.org/eseliger", "username": "eseliger", "avatar_url": "https://secure.gravatar.com/avatar/12bba0958b79429975b4f9a41d709ece?s=80&d=identicon", "identities": null}, "system": true, "created_at": "2020-07-31T11:43:51.215Z"}, {"id": 35199, "body": "closed", "author": {"id": 31, "name": "Erik Seliger", "email": "", "state": "active", "web_url": "https://gitlab.sgdev.org/eseliger", "username": "eseliger", "avatar_url": "https://secure.gravatar.com/avatar/12bba0958b79429975b4f9a41d709ece?s=80&d=identicon", "identities": null}, "system": true, "created_at": "2020-07-31T11:43:42.436Z"}, {"id": 35198, "body": "reopened", "author": {"id": 31, "name": "Erik Seliger", "email": "", "state": "active", "web_url": "https://gitlab.sgdev.org/eseliger", "username": "eseliger", "avatar_url": "https://secure.gravatar.com/avatar/12bba0958b79429975b4f9a41d709ece?s=80&d=identicon", "identities": null}, "system": true, "created_at": "2020-07-31T11:43:16.557Z"}, {"id": 35197, "body": "closed", "author": {"id": 31, "name": "Erik Seliger", "email": "", "state": "active", "web_url": "https://gitlab.sgdev.org/eseliger", "username": "eseliger", "avatar_url": "https://secure.gravatar.com/avatar/12bba0958b79429975b4f9a41d709ece?s=80&d=identicon", "identities": null}, "system": true, "created_at": "2020-07-31T11:43:11.671Z"}, {"id": 35196, "body": "reopened", "author": {"id": 31, "name": "Erik Seliger", "email": "", "state": "active", "web_url": "https://gitlab.sgdev.org/eseliger", "username": "eseliger", "avatar_url": "https://secure.gravatar.com/avatar/12bba0958b79429975b4f9a41d709ece?s=80&d=identicon", "identities": null}, "system": true, "created_at": "2020-07-31T11:43:07.059Z"}, {"id": 35195, "body": "closed", "author": {"id": 31, "name": "Erik Seliger", "email": "", "state": "active", "web_url": "https://gitlab.sgdev.org/eseliger", "username": "eseliger", "avatar_url": "https://secure.gravatar.com/avatar/12bba0958b79429975b4f9a41d709ece?s=80&d=identicon", "identities": null}, "system": true, "created_at": "2020-07-31T11:43:02.771Z"}, {"id": 35194, "body": "reopened", "author": {"id": 31, "name": "Erik Seliger", "email": "", "state": "active", "web_url": "https://gitlab.sgdev.org/eseliger", "username": "eseliger", "avatar_url": "https://secure.gravatar.com/avatar/12bba0958b79429975b4f9a41d709ece?s=80&d=identicon", "identities": null}, "system": true, "created_at": "2020-07-31T11:41:47.619Z"}, {"id": 35193, "body": "closed", "author": {"id": 31, "name": "Erik Seliger", "email": "", "state": "active", "web_url": "https://gitlab.sgdev.org/eseliger", "username": "eseliger", "avatar_url": "https://secure.gravatar.com/avatar/12bba0958b79429975b4f9a41d709ece?s=80&d=identicon", "identities": null}, "system": true, "created_at": "2020-07-31T11:41:39.03Z"}, {"id": 35192, "body": "mentioned in commit 0440892751ca39eb2cdbf88ca2e4e42ea9252cb4", "author": {"id": 31, "name": "Erik Seliger", "email": "", "state": "active", "web_url": "https://gitlab.sgdev.org/eseliger", "username": "eseliger", "avatar_url": "https://secure.gravatar.com/avatar/12bba0958b79429975b4f9a41d709ece?s=80&d=identicon", "identities": null}, "system": true, "created_at": "2020-07-28T16:46:47.826Z"}, {"id": 35191, "body": "merged", "author": {"id": 31, "name": "Erik Seliger", "email": "", "state": "active", "web_url": "https://gitlab.sgdev.org/eseliger", "username": "eseliger", "avatar_url": "https://secure.gravatar.com/avatar/12bba0958b79429975b4f9a41d709ece?s=80&d=identicon", "identities": null}, "system": true, "created_at": "2020-07-28T16:46:46.701Z"}, {"id": 35190, "body": "reopened", "author": {"id": 31, "name": "Erik Seliger", "email": "", "state": "active", "web_url": "https://gitlab.sgdev.org/eseliger", "username": "eseliger", "avatar_url": "https://secure.gravatar.com/avatar/12bba0958b79429975b4f9a41d709ece?s=80&d=identicon", "identities": null}, "system": true, "created_at": "2020-07-28T16:45:53.65Z"}, {"id": 35189, "body": "closed", "author": {"id": 31, "name": "Erik Seliger", "email": "", "state": "active", "web_url": "https://gitlab.sgdev.org/eseliger", "username": "eseliger", "avatar_url": "https://secure.gravatar.com/avatar/12bba0958b79429975b4f9a41d709ece?s=80&d=identicon", "identities": null}, "system": true, "created_at": "2020-07-28T16:44:52.111Z"}, {"id": 35188, "body": "reopened", "author": {"id": 31, "name": "Erik Seliger", "email": "", "state": "active", "web_url": "https://gitlab.sgdev.org/eseliger", "username": "eseliger", "avatar_url": "https://secure.gravatar.com/avatar/12bba0958b79429975b4f9a41d709ece?s=80&d=identicon", "identities": null}, "system": true, "created_at": "2020-07-28T16:44:15.435Z"}, {"id": 35187, "body": "closed", "author": {"id": 31, "name": "Erik Seliger", "email": "", "state": "active", "web_url": "https://gitlab.sgdev.org/eseliger", "username": "eseliger", "avatar_url": "https://secure.gravatar.com/avatar/12bba0958b79429975b4f9a41d709ece?s=80&d=identicon", "identities": null}, "system": true, "created_at": "2020-07-28T16:44:12.901Z"}, {"id": 35186, "body": "reopened", "author": {"id": 31, "name": "Erik Seliger", "email": "", "state": "active", "web_url": "https://gitlab.sgdev.org/eseliger", "username": "eseliger", "avatar_url": "https://secure.gravatar.com/avatar/12bba0958b79429975b4f9a41d709ece?s=80&d=identicon", "identities": null}, "system": true, "created_at": "2020-07-28T16:43:43.554Z"}, {"id": 35185, "body": "closed", "author": {"id": 31, "name": "Erik Seliger", "email": "", "state": "active", "web_url": "https://gitlab.sgdev.org/eseliger", "username": "eseliger", "avatar_url": "https://secure.gravatar.com/avatar/12bba0958b79429975b4f9a41d709ece?s=80&d=identicon", "identities": null}, "system": true, "created_at": "2020-07-28T16:43:29.138Z"}, {"id": 35180, "body": "closed", "author": {"id": 31, "name": "Erik Seliger", "email": "", "state": "active", "web_url": "https://gitlab.sgdev.org/eseliger", "username": "eseliger", "avatar_url": "https://secure.gravatar.com/avatar/12bba0958b79429975b4f9a41d709ece?s=80&d=identicon", "identities": null}, "system": true, "created_at": "2020-07-28T16:26:53.688Z"}, {"id": 35179, "body": "reopened", "author": {"id": 31, "name": "Erik Seliger", "email": "", "state": "active", "web_url": "https://gitlab.sgdev.org/eseliger", "username": "eseliger", "avatar_url": "https://secure.gravatar.com/avatar/12bba0958b79429975b4f9a41d709ece?s=80&d=identicon", "identities": null}, "system": true, "created_at": "2020-07-28T16:19:14.719Z"}, {"id": 35178, "body": "closed", "author": {"id": 31, "name": "Erik Seliger", "email": "", "state": "active", "web_url": "https://gitlab.sgdev.org/eseliger", "username": "eseliger", "avatar_url": "https://secure.gravatar.com/avatar/12bba0958b79429975b4f9a41d709ece?s=80&d=identicon", "identities": null}, "system": true, "created_at": "2020-07-28T16:15:49.104Z"}, {"id": 35177, "body": "unassigned @eseliger", "author": {"id": 31, "name": "Erik Seliger", "email": "", "state": "active", "web_url": "https://gitlab.sgdev.org/eseliger", "username": "eseliger", "avatar_url": "https://secure.gravatar.com/avatar/12bba0958b79429975b4f9a41d709ece?s=80&d=identicon", "identities": null}, "system": true, "created_at": "2020-07-28T16:15:12.329Z"}, {"id": 35176, "body": "assigned to @eseliger", "author": {"id": 31, "name": "Erik Seliger", "email": "", "state": "active", "web_url": "https://gitlab.sgdev.org/eseliger", "username": "eseliger", "avatar_url": "https://secure.gravatar.com/avatar/12bba0958b79429975b4f9a41d709ece?s=80&d=identicon", "identities": null}, "system": true, "created_at": "2020-07-28T16:15:07.689Z"}, {"id": 35175, "body": "changed the description", "author": {"id": 31, "name": "Erik Seliger", "email": "", "state": "active", "web_url": "https://gitlab.sgdev.org/eseliger", "username": "eseliger", "avatar_url": "https://secure.gravatar.com/avatar/12bba0958b79429975b4f9a41d709ece?s=80&d=identicon", "identities": null}, "system": true, "created_at": "2020-07-28T16:15:01.053Z"}, {"id": 35174, "body": "added 7 commits\n\n<ul><li>9b23ff7b...94e4be96 - 6 commits from branch <code>master</code></li><li>c00d24d2 - Merge branch &#39;master&#39; into &#39;kz/release-scripts&#39;</li></ul>\n\n[Compare with previous version](/sourcegraph/bitbucket-server-plugin/merge_requests/19/diffs?diff_id=13672&start_sha=9b23ff7b39d80213bc181645b20b194e5d5392df)", "author": {"id": 31, "name": "Erik Seliger", "email": "", "state": "active", "web_url": "https://gitlab.sgdev.org/eseliger", "username": "eseliger", "avatar_url": "https://secure.gravatar.com/avatar/12bba0958b79429975b4f9a41d709ece?s=80&d=identicon", "identities": null}, "system": true, "created_at": "2020-07-28T16:13:02.088Z"}], "state": "merged", "title": "Kz/upsert webhooks", "labels": [], "web_url": "https://gitlab.sgdev.org/sourcegraph/bitbucket-server-plugin/merge_requests/22", "Pipelines": [{"id": 55, "ref": "kz/release-scripts", "sha": "c00d24d2b0805292f74679b44928f5d20e84166c", "status": "pending", "web_url": "https://gitlab.sgdev.org/sourcegraph/bitbucket-server-plugin/pipelines/55", "created_at": "2020-07-28T16:13:01.04Z", "updated_at": "2020-07-28T16:13:02.394Z"}], "closed_at": null, "diff_refs": {"base_sha": "0440892751ca39eb2cdbf88ca2e4e42ea9252cb4", "head_sha": "d3647d131ff88f2fe3393d1cb20b6de208f39c5c", "start_sha": "0440892751ca39eb2cdbf88ca2e4e42ea9252cb4"}, "merged_at": "2020-07-31T11:48:23.35Z", "created_at": "2020-07-31T11:40:01.898Z", "project_id": 4074, "updated_at": "2020-07-31T11:48:23.214Z", "description": "", "head_pipeline": null, "source_branch": "kz/upsert-webhooks", "target_branch": "master"} | 22 | gitlab |   | kz/upsert-webhooks | 2020-07-31 13:48:23.214+02 | MERGED | PENDING | PENDING | FALSE | TRUE | 0 | 1 | 0 | {"BaseRefOid": "0440892751ca39eb2cdbf88ca2e4e42ea9252cb4", "HeadRefOid": "d3647d131ff88f2fe3393d1cb20b6de208f39c5c", "IsComplete": true} | 1
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --

</details>
<details>
<summary>Changeset event row</summary>

35 | 1 | gitlab:pipeline | Pipeline:55 | 2020-07-31 14:32:43.704077+02 | {"id": 55, "ref": "kz/release-scripts", "sha": "c00d24d2b0805292f74679b44928f5d20e84166c", "status": "pending", "web_url": "https://gitlab.sgdev.org/sourcegraph/bitbucket-server-plugin/pipelines/55", "created_at": "2020-07-28T16:13:01.04Z", "updated_at": "2020-07-28T16:13:02.394Z"} | 2020-07-31 14:32:43.704077+02
-- | -- | -- | -- | -- | -- | --

</details>

LawnGnome commented:
> Yeah I still have it in my database, and for some spurious reason, there actually is _one_ event in the DB:

Thanks!

It's interesting. So there was a pipeline, which correctly created a pending event, and then it got cancelled. I'll investigate why that didn't turn into a more appropriate status on the next sync, if not on a webhook that wasn't recorded.

That said, I don't think this is on the critical path for this PR. I've filed #12667 to fix that issue. Beyond that, if there's no screaming, I'm going to merge this tomorrow morning. (For today is a holiday. I'm not here. This isn't happening.)
